File: cpp/src/lib/checksum.cc
/*
Copyright (C) 2016 iNuron NV

This file is part of Open vStorage Open Source Edition (OSE), as available from


    http://www.openvstorage.org and
    http://www.openvstorage.com.

This file is free software; you can redistribute it and/or modify it
under the terms of the GNU Affero General Public License v3 (GNU AGPLv3)
as published by the Free Software Foundation, in version 3 as it comes
in the <LICENSE.txt> file of the Open vStorage OSE distribution.

Open vStorage is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY of any kind.
*/

#include "checksum.h"

namespace alba {

void Sha1::print(std::ostream &os) const {
  os << "Sha1(`";
  for (char c : _digest) {
    alba::stuff::dump_hex(os, c);
  }
  os << "`)";
}

void Crc32c::print(std::ostream &os) const {
  os << "Crc32c 0x";
  uint32_t d = _digest;
  using namespace alba::stuff;
  dump_hex(os, d >> 24);
  dump_hex(os, (d >> 16) & 0xff);
  dump_hex(os, (d >> 8) & 0xff);
  dump_hex(os, (d)&0xff);
}

bool verify(const Checksum &c0, const Checksum &c1) {
  algo_t a0 = c0.get_algo();
  algo_t a1 = c1.get_algo();
  if (a0 != a1) {
    return false;
  } else {
    switch (a0) {
    case algo_t::NO_CHECKSUM:
      return true;
    case algo_t::SHA1: {
      auto &s0 = (const Sha1 &)c0;
      auto &s1 = (const Sha1 &)c1;
      return (s0._digest == s1._digest);
    }
    case algo_t::CRC32c: {
      auto &s0 = (const Crc32c &)c0;
      auto &s1 = (const Crc32c &)c1;
      return (s0._digest == s1._digest);
    }
    }
  }
  // g++ issues bogus:
  // warning: control reaches end of non-void function [-Wreturn-type]
  return false;
}

std::ostream &operator<<(std::ostream &os, const algo_t &algo) {
  switch (algo) {
  case algo_t::NO_CHECKSUM:
    os << "NO_CHECKSUM";
    break;
  case algo_t::SHA1:
    os << "SHA1";
    break;
  case algo_t::CRC32c:
    os << "Crc32c";
    break;
  }
  return os;
}

std::ostream &operator<<(std::ostream &os, const Checksum &c) {
  c.print(os);
  return os;
}
}

namespace alba {
namespace llio {
template <> void from(message &m, alba::Checksum *&r) {
  uint8_t type;
  from(m, type);
  switch (type) {
  case 1: {
    r = new alba::NoChecksum();
  }; break;
  case 2: {
    std::string digest;
    from(m, digest);
    r = new alba::Sha1(digest);
  }; break;
  case 3: {
    uint32_t d;
    from(m, d);
    r = new alba::Crc32c(d);
  } break;
  default:
    throw "serialization error";
  }
}

template <> void from(message &m, std::unique_ptr<alba::Checksum> &p) {
  alba::Checksum *c;
  from(m, c);
  p.reset(c);
}

template <>
void to(message_builder &mb, alba::Checksum const *const &checksum) noexcept {
  checksum->to(mb);
}
}
}

File: ocaml/src/tools/checksum.ml
(*
Copyright (C) 2016 iNuron NV

This file is part of Open vStorage Open Source Edition (OSE), as available from


    http://www.openvstorage.org and
    http://www.openvstorage.com.

This file is free software; you can redistribute it and/or modify it
under the terms of the GNU Affero General Public License v3 (GNU AGPLv3)
as published by the Free Software Foundation, in version 3 as it comes
in the <LICENSE.txt> file of the Open vStorage OSE distribution.

Open vStorage is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY of any kind.
*)

open Prelude

module Checksum = struct
  module Algo = struct
    type t =
      | NO_CHECKSUM
      | SHA1
      | CRC32c
          [@@deriving show]

    let to_buffer buf = function
      | NO_CHECKSUM -> Llio.int8_to buf 1
      | SHA1 -> Llio.int8_to buf 2
      | CRC32c -> Llio.int8_to buf 3

    let from_buffer buf =
      match Llio.int8_from buf with
      | 1 -> NO_CHECKSUM
      | 2 -> SHA1
      | 3 -> CRC32c
      | k -> Prelude.raise_bad_tag "Checksum.Algo" k
  end

  type algo = Algo.t [@@deriving show]

  type t =
    | NoChecksum
    | Sha1 of HexString.t (* a string of size 20, actually *)
    | Crc32c of HexInt32.t
  [@@deriving show]

  let output buf = function
    | NoChecksum ->
      Llio.int8_to buf 1
    | Sha1 d ->
      Llio.int8_to buf 2;
      assert (String.length d = 20);
      Llio.string_to buf d
    | Crc32c d ->
      Llio.int8_to buf 3;
      Llio.int32_to buf d

  let input buf =
    match Llio.int8_from buf with
    | 1 -> NoChecksum
    | 2 ->
      let d = Llio.string_from buf in
      assert(String.length d = 20);
      Sha1 d
    | 3 ->
      let d = Llio.int32_from buf in
      Crc32c d
    | k -> Prelude.raise_bad_tag "checksum" k

  let deser = input, output
  let from_buffer, to_buffer = input, output

  let algo_of =
    let open Algo in
    function
    | NoChecksum -> NO_CHECKSUM
    | Sha1 _ -> SHA1
    | Crc32c _ -> CRC32c
end
